crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -473,21 +477,110 @@ async fn execute_transactions(
         }
     }
 
-    transaction_manager.enqueue(executable_txns, &epoch_store)?;
+    // Pull out change_epoch tx so that we can special case its execution last. All
+    // of the below block is bookkeeping across the other data structures so that
+    // we can maintain consistent ordering for later assertion checks
+    let (change_epoch_tx, change_epoch_digest, change_epoch_ex_digest) = if end_of_epoch {
+        if executable_txns
+            .last()
+            .expect("End of epoch checkpoint should at least have one tx (change epoch tx)")
+            .transaction_data()
+            .is_change_epoch_tx()
+        {
+            (
+                Some(executable_txns.pop().unwrap()),
+                Some(all_tx_digests.pop().unwrap()),
+                Some(execution_digests.pop().unwrap()),
+            )
+        } else {
+            let index = executable_txns
+                .iter()
+                .position(|tx| tx.transaction_data().is_change_epoch_tx())
+                .expect("End of epoch checkpoint should contain change_epoch tx");
+            (
+                Some(executable_txns.remove(index)),
+                Some(all_tx_digests.remove(index)),
+                Some(execution_digests.remove(index)),
+            )
+        }
+    } else {
+        (None, None, None)
+    };
+
+    let mut effects = execute_transactions_impl(
+        executable_txns,
+        all_tx_digests.clone(),
+        authority_store.clone(),
+        epoch_store.clone(),
+        transaction_manager.clone(),
+        log_timeout_sec,
+    )
+    .await?;
+
+    // As the change_epoch tx is not causally dependent on any other tx,
+    // we must special case it by scheduling it for execution after all other txns
+    // have finished executing
+    if let Some(change_epoch_tx) = change_epoch_tx {
+        let mut change_epoch_fx = execute_transactions_impl(
```

Comment:
What guarantees that at this point all previous transactions have been executed?

Reply:
Oh wow I was working on this late last night and missed a critical point. We could execute the checkpoints out of order. Will fix.

Reply:
Good now. Should be more obviously correct now. We're executing the change_epoch tx now in `check_epoch_last_checkpoint`